src/reconcile/rec_visibility.c

```diff
@@ -357,6 +398,58 @@ __wt_rec_upd_select(WT_SESSION_IMPL *session, WT_RECONCILE *r, WT_INSERT *ins, v
             break;
     }
 
+    /*
+     * Check whether the earliest update in the chain is out-of-order in relation to the current
+     * on-disk value. If the chain is out-of-order, the updates need to be pinned to the cache.
+     *
+     * If the previous update exists on the disk in some form (either because the update has been
+     * restored from the disk OR it has been marked as having already been written to the history
+     * store), then we know that the update has already been made durable so we shouldn't worry
+     * about trying to pin it to the cache.
+     */
+    if (!F_ISSET(r, WT_REC_IN_MEMORY) && prev_upd != NULL && prev_upd->txnid != WT_TXN_ABORTED &&
+      !F_ISSET(prev_upd, WT_UPDATE_HS | WT_UPDATE_RESTORED_FROM_DS | WT_UPDATE_RESTORED_FROM_HS) &&
+      vpack != NULL) {
+        /*
+         * If the on-disk value has a stop time point, then let's compare with the implicit
+         * tombstone that it represents.
+         */
+        if (WT_TIME_WINDOW_HAS_STOP(&vpack->tw)) {
+            cmp_ts = vpack->tw.stop_ts;
+            cmp_txnid = vpack->tw.stop_txn;
+
+            /*
+             * If we had pinned updates last reconciliation, it's possible that the tombstone and
+             * update that form the onpage value may still be in the update list. If that's the
+             * case, we can't expect the last update (which corresponds to the start time point of
+             * the onpage value), to be later than the stop time point. We can easily check if that
+             * is the case by comparing the last update in the chain with the start time point.
+             */
+            if (prev_upd->start_ts == vpack->tw.start_ts &&
```

Comment:
I don't understand this. Why the updates pinned in cache is written to disk?

Reply:
I think it can happen if there was something uncommitted earlier in the chain last time we reconciled. So we'd choose an update for the disk but still pin the list.

Reply:
If the page is update restored, we will free the onpage value. In this case, we will only have the uncommitted tombstone on the update chain.

Reply:
If the page is checkpointed, the in-memory disk image doesn't change.

Reply:
Ok. Well, I saw an update in the chain that corresponded directly to the onpage value. It didn't have any flags either so it can't be because of appending.

Reply:
Currently, how eviction works is if something is written to the disk, from the onpage value, everything is freed from memory. The only exception is the onpage tombstone which is newer to the onpage value.

Reply:
I think what you describe should only happen in in-memory case. If it happens in mongodb, I think there may be something wrong there.

Reply:
Hmm, maybe this was before I added the check for the in-memory case.
I'll see if I still spot that issue when I remove the check in the test suite.

Reply:
Yeah, you're correct. I removed this check and it seems ok. I think I added this code in response to an in-memory test failure.